R/module_teal_lockfile.R

```diff
@@ -0,0 +1,212 @@
+        result <- process$result()
+        if (any(grepl("Lockfile written to", result$out))) {
+          logger::log_debug("Lockfile containing { length(result$res$Packages) } packages created.")
+          if (any(grepl("(WARNING|ERROR):", result$out))) {
+            warning("Lockfile created with warning(s) or error(s):", call. = FALSE)
+            for (i in result$out) {
+              warning(i, call. = FALSE)
+            }
+          }
+          enable_lockfile_download()
+        } else {
+          disable_lockfile_download()
+        }
+      } else if (process$status() == "error") {
+        disable_lockfile_download()
+      }
+    })
+
+    NULL
+  })
+}
+
+utils::globalVariables(c("opts", "sysenv", "libpaths", "wd", "lockfilepath", "run")) # needed for mirai call
+#' @rdname module_teal_lockfile
+.teal_lockfile_process_invoke <- function(lockfile_path) {
+  mirai_obj <- NULL
+  process <- shiny::ExtendedTask$new(function() {
+    m <- mirai::mirai(
+      {
+        options(opts)
+        do.call(Sys.setenv, sysenv)
+        .libPaths(libpaths)
+        setwd(wd)
+        run(lockfile_path = lockfile_path)
+      },
+      run = .renv_snapshot,
+      lockfile_path = lockfile_path,
+      opts = options(),
+      libpaths = .libPaths(),
+      sysenv = as.list(Sys.getenv()),
+      wd = getwd()
+    )
+    mirai_obj <<- m
+    m
+  })
+
+  shiny::onStop(function() {
+    if (mirai::unresolved(mirai_obj)) {
+      logger::log_debug("Terminating a running lockfile process...")
+      mirai::stop_mirai(mirai_obj) # this doesn't stop running - renv will be created even if session is closed
+    }
+  })
+
+  suppressWarnings({ # 'package:stats' may not be available when loading
+    process$invoke()
+  })
+
+  logger::log_debug("Lockfile creation started based on { getwd() }.")
+
+  process
+}
+
+#' @rdname module_teal_lockfile
+.renv_snapshot <- function(lockfile_path) {
+  out <- utils::capture.output(
+    res <- renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE,
+      type = renv::settings$snapshot.type() # see the section "Different ways of creating lockfile" above here
+    )
+  )
+
+  list(out = out, res = res)
+}
+
+#' @rdname module_teal_lockfile
+.is_lockfile_deps_installed <- function() {
+  requireNamespace("mirai", quietly = TRUE) && requireNamespace("renv", quietly = TRUE)
+}
+
+#' @rdname module_teal_lockfile
+.is_lockfile_enabled <- function() {
+  if (isTRUE(getOption("teal.lockfile.enable"))) {
+    return(TRUE)
+  } else {
+    !(
+      identical(Sys.getenv("CALLR_IS_RUNNING"), "true") || # inside callr process
+        identical(Sys.getenv("TESTTHAT"), "true") || # inside devtools::test
+        !identical(Sys.getenv("QUARTO_PROJECT_ROOT"), "") || # inside Quarto process
+        (
+          ("CheckExEnv" %in% search()) || any(c("_R_CHECK_TIMINGS_", "_R_CHECK_LICENSE_") %in% names(Sys.getenv()))
+        ) # inside R CMD CHECK
+    )
+  }
+}
```

Comment:
I have some concerns on how the `teal.lockfile.enable` option is used. 
- This is not set by us and it would probably be never set by the end-users.
- If I set it to FALSE then (assuming all conditions are not met) I still might end up with lockfile created

How about below: three possible option values: true, auto, false and on load we set it to auto.

Reply:
@pawelru good point

```r
if (<is set>) {
  getOption("teal.lockfile.enable")
} else {
  ...
}
```

Reply:
Some additional thoughts:

- `"enabled"` option name typically denotes boolean with true/false choices so from that perspective it would make sense to _merge_ the "true" and "auto" value

- if we want to continue with this we can have the following values: auto, user, false. That option should be named more as a "mode" instead

- why do we need the true value? I think we are more interested in false instead. If that's for testing then the answer is to temporarily mock that `.is_lockfile_enabled()` function (and not reverse engineer options values)

Reply:
It make sense to 
```
if (<option is set>)
  getOption("teal.lockfile.enable")
} else {
```


Just saying (and it is a general comment not specifically about this solution) that setting option in the parent process doesn't always guarantee that all children will preserve an option. It depends on a child process (e.g. `run_examples(fresh = TRUE)` function `build_vignettes()`) passes options further.

Reply:
So this option is definitely needed to allow for explicit user disabling if one does not want it. So we need the "false" value. 
I just can't imagine a user (!) scenario where the "true" value would be used. This value would indicate more or less the following: ignore all of those checks that we came up with and create lockfile no matter what. These checks suppose to eliminate scenarios that occurs only within package development. A "standard" usage should not be covered by these. On the other hand - we can make mistakes in our checks so this mechanism opens the doors for overwrittes. So maybe it's not that bad.
I can image a scenario where "true" is needed - our tests. But like I said earlier - if we have that logic niceley encapsulated in one function `.is_lockfile_enabled()` - we should mock that fun instead. 

Re interdependent options - I started to like `user` option value. Currently we have a following logic: if user provides a file path and such file does not exist (or inaccessible) then fallback to another option value checks. This is probably a little bit too convoluted and also difficult to implement and document. A "user" option value (alongside another path option value) makes it clear that I want user provided file (and don't have fallback mechanism). We can easily return from [here](https://github.com/insightsengineering/teal/blob/f76a30ff4403e45e5ff18fcc9ccea05b84d00d72/R/module_teal_lockfile.R#L104-L106). 
I believe that this would make it more explicit and simpler. Especially for the user.

Reply:
> Re interdependent options - I started to like user option value. Currently we have a following logic: if user provides a file path and such file does not exist (or inaccessible) then fallback to another option value checks. This is probably a little bit too convoluted and also difficult to implement and document. A "user" option value (alongside another path option value) makes it clear that I want user provided file (and don't have fallback mechanism). We can easily return from [here](https://github.com/insightsengineering/teal/blob/f76a30ff4403e45e5ff18fcc9ccea05b84d00d72/R/module_teal_lockfile.R#L104-L106).
I believe that this would make it more explicit and simpler. Especially for the user.

Totally agree. We had a similar discussion in here https://github.com/insightsengineering/teal/pull/1263#discussion_r1760862414

So I guess it should/could be
- `teal.lockfile.path` 
    - when provided, it's superior to `teal.lockfile.enable` and if the file exists then it's possible to download the pre-computed file
    - when provided, but file does not exist, we `return(NULL)` from [here](https://github.com/insightsengineering/teal/blob/f76a30ff4403e45e5ff18fcc9ccea05b84d00d72/R/module_teal_lockfile.R#L104-L106) and we do not compute lockfile
    - when not provided, then regular flow for lockfile creation is followed
    
Regular flow, when `teal.lockfile.path` not provided
- check `teal.lockfile.emabled`
    - when set to TRUE, compute anyway (case for tests in our scenario, and unpredicted user scenarios)
    - when set to FALSE, do not compute
    - when not set - check IF_QUATRO, IF_TESTING, IF_R_CMD_CHECK, IS_CALLR 
        - if TRUE, omit
        - if FALSE, proceed
        
        
For the documentation - let's make this behavior clear in `options-vignette`. Let's link to `options-vignette` in the manual page of `module_teal_lockfile`


How does that sound @pawelru  :) ?

Reply:
I am moving towards replacing `.enabled` option with a new one `.mode` which is checked first.

The logic would be as follows:
```mermaid
flowchart TD
    A{teal.lockfile.mode}
    A -->|true| yes[yes]
    A -->|auto| B{check conditions}
    B -->|true| yes
    B -->|false| B_no[logger]
    B_no --> no[no]
    A -->|user| C{teal.lockfile.path}
    A -->|false| A_false[logger]
    A_false --> no
    C -->|empty| C_empty[logger]
    C_empty --> no
    C -->|non-empty| F[use what's provided]
```

The biggest difference is to eliminate the "if smth is provided then its superior to smth" and "if condition is not met then fallback to regular". These are the weakest part IMHO.

Reply:
Oh my, this graph doesn't make me think anything got simplified : P

Reply:
So `teal.lockfile.mode` can be `true`, `false`, `auto` and `user`. And then if `teal.lockfile.mode = 'user'` we also check if `teal.lockfile.path` is empty, non-empty and exists, non-empty and does not exist.

I think it's harder to understand with 4 `options` for `mode` and 3 `scenarios` for `path` and harder to document. 

> The biggest difference is to eliminate the "if smth is provided then its superior to smth"

But it also exists in your flow, if `teal.lockfile.mode` it is provided it is superior to `teal.lockfile.path`.
I think with two options we will always need to prioritize one of it.

Other idea
1 option `teal.lockfile.mode` with values: `true`, `false`, `auto` and `actual_path`.

Reply:
> > The biggest difference is to eliminate the "if smth is provided then its superior to smth"
> 
> But it also exists in your flow, if `teal.lockfile.mode` it is provided it is superior to `teal.lockfile.path`. I think with two options we will always need to prioritize one of it.

There is a slight difference. What I am suggesting is to have one option to be always (!) superior than the other one. I want to eliminate auto-magic: if one is provided then it's superior.
The usage of `.path` is not conditional on its existence but rather based on another `.mode` value.

The second biggest difference to your (previous) suggestion is that if we got the "user" mode turned on then we go hard with that mode and throw if cannot be done - in particular: do not fallback to the auto mode.
I do consider "user" mode together with a incorrect path as a wrong input and I would like to notify user about that.

> Other idea
> 1 option teal.lockfile.mode with values: true, false, auto and actual_path.

What if my lockfile path is `true` or `false` or `auto`? I think it's a mistake to merge responsibilities of two vars into one. One responsibility is "mode" choice - one out of fixed scope of available choices. The other one is a path. The latter is used only if the former has some specific value.

Reply:
I would assume your lockfile path has `.lock` at the end, so it would rather not be `true`, `false` nor `auto` but rather `true.lock`, `false.lock`, or `auto.lock`. If `true.lock` or `auto.lock` are way to popular names and someone can name their lockfiles without extensions and use `true` or `auto`, then we can change mode to use `disabled` and `enabled`.

My point is that if `teal.lockfile.mode` is already a character, then it could take the path. And maybe we don't need 2 options. We can put a condition so that path needs to be specified with a prefixed tag: `option(teal.lockfile.mode = "path:custom_user_path.lock")`. Then you have a clear distinction on whether this is an actual path or type of the mode. 

```r
grepl('^path:', getOption('teal.lockfile.mode')) 
```

Let's you know if this is user provided lockfile, or all other possibilities

Reply:
> This is an interesting idea. I already have seen this in pkgdepends to install packages from different sources so you are giving the reference and the category of that reference which is a string, e.g. git::some_url or github::some_url or gitlab::some_url or local::path. So it's a way to classify / categorise string type of data which is an obligatory input.

Yes, I've seen it in `pkgdepends` :)

>  I think it's easier to have this as a separate being.

Alrighty, then let's keep it as you proposed. I'm on board!

- Check `teal.lockfile.mode` and if:
    - `auto` set on the package startup, does not look at `teal.lockfile.path`, goes into SCENARIO conditions
    - `true`, need to be set manually, does not look at `teal.lockfile.path`, does not goes into SCENARIO conditions, just computes
    - `false`, need to be set manually, does not look at `teal.lockfile.path`, does not goes into SCENARIO conditions, does not compute
    - `user`, need to be set manually, does look at `teal.lockfile.path` - stops when does not exist or empty, computes when not-empty and exists
    
Each path that ends up with lockfile not being computed shows logger trace.

Sounds good

Reply:
👍

Reply:
I don't like `teal.lockfile.mode` option to have 4 levels and `teal.lockfile.path` being considered only when `.mode == "user"`. I'm not a fan of specifying two options in order to provide pre-calculated file. I like the idea of `teal.lockfile.path` to automatically enable lockfile. Maybe the problem is in options naming? I think these options names are more relevant.

- `teal.lockfile.user` - If specified, lockfile is read from the given location. Has a precendence over `teal.lockfile.auto`
- `teal.lockfile.auto` (or `teal.lockfile.system`) - when set to `TRUE` (conditoinal default) then lockfile is calculated automatically using home directory.

<img width="821" alt="image" src="https://github.com/user-attachments/assets/c9bc7781-db7d-408e-af3f-b85f05406b37">

Of course we need a initial warning if `teal.lockfile.auto = TRUE` and `teal.lockfile.user = "<not empty>` saying "teal.lockfile.user has a precedence over teal.lockfile.auto. Reading a lockfile from the user specified location"

Reply:
Just the summary of the discussion we had today - we remove `user` option, hence `option('teal.lockfile.path')` as it is risky to allow to display custom renv.lockfile that might be different than the actual lockfile of the project (potentially outdated). We will proceed with just 3 options: `disable`, `enable`, `auto`